### PR TITLE
(HI-328) Prevent interpolation recursion

### PR DESCRIPTION
**Pull request by thallgren**
This commit ensures that only one RecursiveGuard instance is
used in a lookup even if that lookup results in multiple
interpolations.

A backend implementation must add and pass on a fifth
'recurse_guard' argument but the implementation will accept
the old API with four arguments as well for backward
compatibility.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,7 +6,7 @@ vendor/
 Gemfile.lock
 Gemfile.local
 /.project
-
 acceptance_tests/.bundle
 acceptance_tests/junit
 acceptance_tests/log
+.idea/
```
Comment (hlindberg) [inline .gitignore]: this is not tied to the HI.238 issue - should really be in a separate commit

File: lib/hiera/backend.rb [hunk 2]
```diff
@@ -196,14 +196,20 @@ def merge_answer(left,right)
...
 
         Config[:backends].each do |backend|
           if constants.include?("#{backend.capitalize}_backend") || constants.include?("#{backend.capitalize}_backend".to_sym)
             @backends[backend] ||= Backend.const_get("#{backend.capitalize}_backend").new
-            new_answer = @backends[backend].lookup(key, scope, order_override, resolution_type)
+            backend = @backends[backend]
+            begin
+              new_answer = backend.lookup(key, scope, order_override, resolution_type, recurse_guard)
+            rescue ArgumentError
+              # Handle old style backend that take four arguments
```
Comment (hlindberg) [inline lib/hiera/backend.rb]: Hm, this means it will catch argument errors that surface from an existing 5 arg lookup as well. I think it needs to look at the arity of the called lookup method instead. It also needs to issue a debug message when the 4 argument version is called. (Not sure if there is a way to issue a deprecation warning in hiera code base (that does outputs the warning once as opposed to on each call to lookup).

Comment (hlindberg) [inline lib/hiera/backend.rb]: i.e. do the patching up front to avoid having to check on each call.

**Comment by hlindberg** [comment]
When merged, tickets should be filed to remove the deprecation of lookup method without guard, and to remove the support for patching older backends. (I.e. target Hiera 3.0)

File: lib/hiera/backend.rb [hunk 3]
```diff
@@ -8,6 +8,17 @@
 
 class Hiera
   module Backend
+    class LegacyBackendWrapper
```
Comment (hlindberg) [inline lib/hiera/backend.rb]: This code is Legacy tomorrow - better to use the version. This class adapts a Hiera 1x backend to the Hiera 2x API. So... Backend1xAdapter ?

File: lib/hiera/backend.rb [hunk 4]
```diff
@@ -8,6 +8,17 @@
...
+      def initialize(wrapped)
+        @wrapped = wrapped
+      end
+
+      def lookup(key, scope, order_override, resolution_type, recurse_guard)
```
Comment (hlindberg) [inline lib/hiera/backend.rb]: is lookup the only method in the API?

**Comment by hlindberg** [comment]
Before merging this, it is prudent to test that Puppet master and this PR works correctly together (rather than waiting until after the release of hiera 2.0.0 and upgrade of puppet to run against this version of hiera. (i.e. run puppet's spec tests against puppet and hiera source)

File: lib/hiera/backend.rb (hunk 4 above)
Comment (thallgren) [inline lib/hiera/backend.rb]: Yes. This is the only method implemented by the backends and they don't inherit, include or extend anything.

**Comment by thallgren** [comment]
I have a branch of puppet/master where I have verified that puppet's spec tests run OK. No PR yet of course since it relies on adjacent hiera source and hence would fail the Travis tests.

https://github.com/thallgren/puppet/tree/issue/hi-328/fix-recursive-guard

**Comment by hlindberg** [comment]
Thanks @thallgren that is great. Create a ticket for Puppet "Update puppet to hiera 2.0.0" targeting PUP 4.0.0 and add the PR there).

File: lib/hiera/backend.rb (hunk 4 above)
Comment (hlindberg) [inline lib/hiera/backend.rb]: ok - this is fine then.
